hpcrawler.py

```diff
@@ -0,0 +1,364 @@
+            if dup_file:
+                os.remove(os.path.join(self._path, filename))  # delete duplicate
+                filename = re.sub(r'\(\d+\)\.gpx', '', filename) + ".gpx"  # work with the original one
+            os.rename(os.path.join(self._path, filename), j_gpx_path + ".gpx")
+
+        else:
+            filename = progress[self._track_idx][0]
+
+        # update and save the progress, what we've mined so far:
+        # NOTE:
+        # (1) we want to save filename for future use: when re-mining a country that
+        # was mined before, we'll be able to see which tracks we've processed before.
+        # (2) we rather save the gpx files by their <j> because it's uniqe and it allows
+        # us to keep mining from where we've stopped (ints are ordered).
+        progress.update({self._track_idx: [filename, track_len, track_dif]})
+        self._save_dict(progress, os.path.join(self._path, "progress.json"))
+
+        return [filename, track_len, track_dif]
+
+# processes data #
+    def _process_track_data(self, features):
+        """
+        process the track's data that was collected.
+        :param features:  the track's features: [filename, track_len, track_dif]
+        :return: the track's len_tag: see SlopeMap
+                the track's representation:  dict {<country>_<j>: [slopes, track_dif]}
+        """
+        # print("\t* processing track") #TODO
+        points, track_elev = pd.DataFrame({}), pd.DataFrame({})
+        filename, track_len, track_dif = features
+        track_len = float(track_len)
+
+        # open gpx file:
+        gpx_file = open(os.path.join(HpCrawler.gpx_dir_path, self._country, self._track_idx + ".gpx"),
+                        'r', encoding="utf8")
+        gpx = gpxpy.parse(gpx_file)
+
+        # get points array & elevations:
+        for track in gpx.tracks:
+            for seg in track.segments:
+                points = pd.DataFrame([{'lat': p.latitude, 'lon': p.longitude} for p in seg.points]) \
+                    .to_numpy()
+                track_elev = pd.DataFrame([{'ele': p.elevation} for p in seg.points]) \
+                    .to_numpy().reshape(len(points))
+
+        # compute slopes:
+        tick = sm.get_tick(track_len)
+        slopes = sm.compute_slope(points, track_elev, tick)
+
+        # save track to the correct dict (according to the track's length):
+        len_tag = sm.get_length_tag(track_len)
+
+        return len_tag, {self._country + '_' + self._track_idx: [slopes, track_dif]}
+
+# runs functionality:
+
+    def crawl(self):
+        """
+        crawls the trails data from "the hiking project"'s site.
+        it saves the data and the progress of crawling under the "hp" directory.
+        :return all of the countries we've crawled so far (not just in this iteration)
+        """
+        for i in range(len(self._countries)):
+            seen = HpCrawler.load_seen()
+            self._country = self._countries[i]
```

Comment:
not sure why _country is a parameter of the object since this is very much related to the process being performed at the moment (I see it used in a lot of places but I think it should be moved as a parameter since its currently unclear what it does without reading the entire code)